Add Minkowski distance metric support to corner_peak

rfezzani opened the pull request:
## Description

`scipy.spatial.distance.pdist` is used to compute the distance between each peak. This allows to use two different metrics: `chebyshev` (corresponding to the actual behavior) and `euclidean`.

This PR manages the  issue #4200.

It can also allow to support non integral values for the `min_distance` parameter, but this may be discussed later.

### Update

Thanks to @emmanuelle suggestion, `scipy.spatial.cKDtree` is used instead of `pdist`. All the range of Minlowski p-distances can be used now (p=inf -> Chebychev; p=2 -> Euclidean...)
Performances are improved and memory is saved.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

pep8speaks commented:
Hello @rfezzani! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/feature/corner.py`](https://github.com/scikit-image/scikit-image/blob/f9156ac04af520873f57c226dbb95e560c4274ac/skimage/feature/corner.py):

> [Line 920:1](https://github.com/scikit-image/scikit-image/blob/f9156ac04af520873f57c226dbb95e560c4274ac/skimage/feature/corner.py#L920): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 920:1](https://github.com/scikit-image/scikit-image/blob/f9156ac04af520873f57c226dbb95e560c4274ac/skimage/feature/corner.py#L920): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 920:4](https://github.com/scikit-image/scikit-image/blob/f9156ac04af520873f57c226dbb95e560c4274ac/skimage/feature/corner.py#L920): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

emmanuelle commented:
Hi @rfezzani sorry it took so long to review the PR! I have some concerns about the performance of the proposed solution: the execution time of the `pdist` function scales like the square of the number of rows. For 10 000 points in 3D I measured 240 ms just for `pdist` on my machine, and for 100 000 points I got a memory error (it sounds like a lot, but for noisy 3d images it is not unrealistic). 

For adding Euclidean distance, which is useful I think, I would recommend instead to use a cKDTree  https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.html which allows you to query points within a certain radius (`query_ball_point`) in linear time. What would you think of this approach?

rfezzani commented:
Excellent suggestion @emmanuelle, I will give it at try ;-)

rfezzani commented:
Hi @emmanuelle, your suggestion works perfectly well :+1:, The code is even more readable now ;-)

jni commented on skimage/feature/corner.py:
```diff
@@ -954,36 +974,48 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
+                            num_peaks_per_label=num_peaks_per_label)
+
+    if len(coords):
+        # Use KDtree to find the peaks that are too close to each others
```
```suggestion
        # Use KDtree to find the peaks that are too close to each other
```

jni commented on skimage/feature/corner.py:
```diff
@@ -954,36 +974,48 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
+                candidates.remove(idx)
+                rejected_peaks.update(candidates)
+
+        # Remove the peaks that are too close to each others
```
```suggestion
        # Remove the peaks that are too close to each other
```

jni commented on skimage/feature/corner.py:
```diff
@@ -954,36 +974,48 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
+                rejected_peaks.update(candidates)
+
+        # Remove the peaks that are too close to each others
+        coords = np.delete(coords, tuple(rejected_peaks), axis=0)[::-1]
```
Why are you inverting the coordinates here? Are they originally in decreasing order of height, btw?

rfezzani replied:
They are originally in decreasing order (which is good), but that reordering is necessary for tests backward compatibility :-(

rfezzani commented:
@jni, do you think that decreasing order should be kept? I would be in favor of this but this will brake backward compatibility and some tests results have to be modified accordingly.
Is then a deprecation cycle necessary here?

rfezzani commented:
To be noted that the previous version was not taking into account the value of the peaks before removal...

jni commented:
@rfezzani yes, I'm in favour of decreasing order, I think it makes the most sense. @scikit-image/core anyone else want to chime in?

emmanuelle commented on skimage/feature/corner.py:
```diff
@@ -915,16 +916,35 @@ def corner_subpix(image, corners, window_size=11, alpha=0.99):
+        Which Minkowski p-norm to use. Should be in the range [1, inf].
+        A finite large p may cause a ValueError if overflow can occur.
+        inf corresponds to the Chebyshev distance and 2 to the
+        euclidean distance.
```
```suggestion
        Euclidean distance.
```

emmanuelle commented:
@jni @rfezzani I'm also in favor of decreasing order. We never guaranteed anything about the order so I would not call this a breaking change, all the more than the new behaviour makes more sense.

alexdesiqueira commented on skimage/feature/corner.py:
```diff
@@ -915,16 +916,35 @@ def corner_subpix(image, corners, window_size=11, alpha=0.99):
+        to set the Minkowski p-norm defining the distance.
     * : *
         See :py:meth:`skimage.feature.peak_local_max`.
+    p : float
```
Could we call this `p` as `p_norm` instead? I think we would benefit from a more descriptive name here.

alexdesiqueira commented on skimage/feature/corner.py:
```diff
@@ -915,16 +916,35 @@ def corner_subpix(image, corners, window_size=11, alpha=0.99):
+    output : ndarray or ndarray of bools
+
+        * If `indices = True`  : (row, column, ...) coordinates of peaks.
+        * If `indices = False` : Boolean array shaped like `image`, with peaks
```
I think these `* If (...)` would be better in a `Notes` section... is it just me?

rfezzani replied:
I personally have no strong opinion on this, but I don't know what to put in `output` description then...

rfezzani replied:
Btw, we are in the 3. of https://github.com/scikit-image/scikit-image/issues/3892#issuecomment-591819705 case here ;-)

rfezzani replied:
@alexdesiqueira, I see here that I simply copied the Return section of [`peak_local_max`](https://scikit-image.org/docs/dev/api/skimage.feature.html#skimage.feature.peak_local_max).

stefanv replied:
I'd probably leave it here, since it describes the return; in the notes section, we typically discuss implementation details that users may be curious about, or that would help them better understand the algorithm.

rfezzani commented:
> @jni @rfezzani I'm also in favor of decreasing order. We never guaranteed anything about the order so I would not call this a breaking change, all the more than the new behaviour makes more sense.

@jni @emmanuelle, `corner_peak` is used in [`feature.ORB`](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/orb.py#L142), so modifying the order will introduce breaking changes...

jni commented on skimage/feature/tests/test_corner.py:
```diff
@@ -1,30 +1,31 @@
+                      corner_fast, corner_orientations,
+                      structure_tensor, structure_tensor_eigvals,
+                      hessian_matrix, hessian_matrix_eigvals,
+                      hessian_matrix_det, shape_index)
```
Typically in our tests we use absolute, not relative imports. Could you revert these changes?

jni commented on skimage/feature/tests/test_brief.py:
```diff
@@ -42,20 +42,20 @@ def test_uniform_mode():
+                         [1, 1, 1, 1, 1, 1, 0, 0],
+                         [1, 1, 1, 1, 1, 1, 1, 1],
+                         [0, 0, 0, 1, 1, 1, 1, 1],
+                         [0, 1, 0, 1, 0, 1, 1, 1]], dtype=bool)
```
Love this. =D

jni commented on skimage/feature/tests/test_brief.py:
```diff
@@ -42,20 +42,20 @@ def test_uniform_mode():
 
     keypoints = corner_peaks(corner_harris(img), min_distance=5,
-                             threshold_abs=0, threshold_rel=0.1)
+                             threshold_abs=0, threshold_rel=0.1)[::-1]
```
Why is the order reversed here?

jni commented on skimage/feature/tests/test_corner.py:
```diff
@@ -361,13 +361,33 @@ def test_corner_peaks():
+                           indices=False, threshold_rel=0.1)
     assert np.sum(corners) == 5
 
+    # Test Euclidean distance
```
Could this be separated into a new test? Tests should be as granular as possible so that just the name of a failed test gives as much information as possible about the failure.

jni commented on skimage/feature/tests/test_corner.py:
```diff
@@ -467,7 +487,7 @@ def test_corner_orientations_astronaut():
                          -4.40598471e-01,  3.14918803e-01, -1.76069982e+00,
                           3.05330950e+00,  2.39291733e+00, -1.22091334e-01,
-                         -3.09279990e-01,  1.45931342e+00])
+                         -3.09279990e-01,  1.45931342e+00])[::-1]
```
I don't understand why orders are reversed... I get that the order has changed, but why has it changed to the exact opposite?

rfezzani commented:
> (a) how come the order is exactly reversed from the existing order in so many tests? I thought originally they were in order of position in the image, while now they are in decreasing order of size. Am I incorrect in that assessment?

@jni, thank you for raising this, I fixed the tests without care... :pray:
After some investigations, I finally found the reason why: The peaks are returned in decreasing order only if their number exceeds the argument `num_peaks` value!
I also noticed that the coordinates are reordered up side down in [`peak_local_max`](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/peak.py#L21). This makes `peak_local_max` returns the coordinates in reverse order compared to `corner_peak`:
```python
>>> import skimage
>>> skimage.__version__
'0.16.2'
>>> from skimage.feature import peak_local_max, corner_peaks
>>> from skimage import data
>>> img = data.coins()
>>> hc = corner_harris(img)
>>> kp = corner_peaks(hc, min_distance=5, threshold_abs=0, threshold_rel=0.1)
>>> lm = peak_local_max(hc, min_distance=5, threshold_abs=0, threshold_rel=0.1)
>>> np.all(kp == lm[::-1])
True
```
 By using directly the coordinates returned by `peak_local_max` (and not the mask), the new implementation returns the coordinates in the reverse order then the original implementation...
I will come with a PR fixing `peak_local_max` coordinates reordering...

rfezzani commented:
> I will come with a PR fixing `peak_local_max` coordinates reordering...

@jni, please see #4501 ;-)

rfezzani commented:
@jni, I fixed peaks coordinates ordering after merging #4501. I also fixed a bug in my implementation: the number of returned peaks was bound by but not equal to `num_peaks`.

jni commented:
@rfezzani it looks like this needs a rebase/merge with master.

rfezzani commented:
@jni, rebase done ;)

emmanuelle commented on skimage/feature/corner.py:
```diff
@@ -917,16 +918,35 @@ def corner_subpix(image, corners, window_size=11, alpha=0.99):
+        The minimum distance seperating peaks. Use the ``p`` argument
+        to set the Minkowski p-norm defining the distance.
     * : *
         See :py:meth:`skimage.feature.peak_local_max`.
```
Should this line be removed? I think so, it was here just because we hadn't written the Parameters section.

rfezzani replied:
Should I copy the doc from `peak_local_max`?

emmanuelle replied:
oh sorry I did not understand that we still had undocumented arguments. No keep it like this then, I will mark the conversation as resolved.

emmanuelle commented on skimage/feature/corner.py:
```diff
@@ -956,38 +976,52 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
+        # Use KDtree to find the peaks that are too close to each other
+        tree = spatial.cKDTree(coords)
+
+        rejected_peaks = set()
```
I would call this variable `rejected_peaks_indices` to make clear that this is a set of indices, not the coordinates themselves.

emmanuelle commented:
So the order of returned peaks changed. Should we document this order? Is this by decreasing amplitude of peaks?

emmanuelle commented:
Ideally the order should be by decreasing amplitude and for the same amplitude from top to bottom and left to right (or for more dims by increasing coordinates)

emmanuelle commented:
I just have a comment about the order of returned peaks but I'm very close to approve this PR :-).

rfezzani commented:
Thank you @emmanuelle for your comments.
> Should we document this order?

We can add note on this if you wish.
> Ideally the order should be by decreasing amplitude and for the same amplitude from top to bottom and left to right (or for more dims by increasing coordinates)

It is the case ;-)

emmanuelle commented:
> It is the case ;-)

Is it? In the doctest example in the docstring or `corner_harris`, all peaks should have the same amplitude and the corners with largest row index come first...

rfezzani commented:
@emmanuelle, you catched a bug here :clap:: to obtain the decreasing order, I choosed in #4501
to revert the indices returned by `argsort` [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/peak.py#L15):
```python
    idx_maxsort = np.argsort(intensities)[::-1]
```
For managing peaks with same amplitude it should be:
```python
    idx_maxsort = np.argsort(-intensities)
```
I will come with the PR ;-)

lagru commented on skimage/feature/corner.py:
```diff
@@ -956,38 +976,53 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
+                            footprint=footprint, labels=labels,
+                            num_peaks_per_label=num_peaks_per_label)
+
+    if len(coords):
+        # Use KDtree to find the peaks that are too close to each other
+        tree = spatial.cKDTree(coords)
+
+        rejected_peaks_indices = set()
+        for idx, point in enumerate(coords):
+            if idx not in rejected_peaks_indices:
+                candidates = tree.query_ball_point(point, r=min_distance,
+                                                   p=p_norm)
+                candidates.remove(idx)
+                rejected_peaks_indices.update(candidates)
+
+        # Remove the peaks that are too close to each other
+        coords = np.delete(coords, tuple(rejected_peaks_indices),
+                           axis=0)[:num_peaks]
```
Sorry to chime in this late in the discussion. I think this may be closely related to #4165. As I understand the distance metric aspect here the output `coords` of `peak_local_max` is a coordinate array of the corner peaks which are then filtered with the help of cKDTree. If that is the case the proposed function `remove_near_objects` in #4165 should be able to replace this part. 

Performance wise I'm not sure which solution would be most appropriate here:
- `remove_near_objects` would require a temporary conversion of `coords` to a boolean mask (`indices=False`) and comes with additional overhead which is not needed for this use case.
- but its speed should scale significantly better with the number of peaks because most of the iteration is done in Cython. 

It could very well be that the break-even point is way beyond what is considered an expected number of peaks here so take this suggestion with a grain of salt until I or someone else had the chance to benchmark this. 

And in any case this shouldn't hold up this PR as it could be done afterwards without deprecations (if we manage to deal with the ordering of the peaks). The way I see it this is already an improvement over the old state. :+1:

rfezzani replied:
I also did not found time to review your PR #4165, so don't worry ;-)
You are right

lagru replied:
Yes, the order by which objects (in this case corner peaks) are removed is determined by the `priority` attribute. It defaults to the sample count of objects (which in this case should be 1 for every object?) but it would be trivial to pass in an array with the correct amplitude values. That would be the "corner measure response image" correct?

stefanv replied:
Even if this cannot utilize another function, it should probably be factored out to shared utilities: it seems very useful!

lagru replied:
In any case something for another PR I'd say.

lagru commented on skimage/feature/corner.py:
```diff
@@ -956,38 +976,53 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
+
+    if len(coords):
+        # Use KDtree to find the peaks that are too close to each other
+        tree = spatial.cKDTree(coords)
```
@rfezzani  Have you played around with the `balanced_tree=True` parameter? Constructing a balanced tree takes more time but should pay of the more we queries we use on it (which should be quite a lot in this case).

rfezzani replied:
I did not, but I definitely will. I will come back after some more tests ;-)

rfezzani replied:
@lagru, I made some more investigation, and I observed no influence of the `balanced_tree` parameter on the computation time:

```python
In [1]: from skimage.feature import corner_peaks, corner_fast                                                                           

In [2]: from skimage.color import rgb2gray                                                                                              

In [3]: from skimage.data import astronaut                                                                                              

In [4]: img =rgb2gray(astronaut())                                                                                                      

In [5]: corners = corner_fast(img, 12, 0.3)                                                                                             

In [6]: %timeit corner_peaks(corners, threshold_rel=0.1) 
```

#### `balanced_tree=True`

```python                                                                
11.6 ms ± 20.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

#### `balanced_tree=False`

```python                                                               
11.5 ms ± 49.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

lagru replied:
Thanks for investigating!

lagru commented on skimage/feature/tests/test_corner.py:
```diff
@@ -128,8 +130,7 @@ def test_hessian_matrix_eigvals_3d(im3d):
     row_center, col_center = np.array(E0.shape) // 2
-    circles = [draw.circle_perimeter(row_center, col_center, radius,
-                                     shape=E0.shape)
+    circles = [circle_perimeter(row_center, col_center, radius, shape=E0.shape)
```
And a small suggestion: you changed the imports in `test_corner.py`, e.g. `skimage import draw` to `from skimage.draw import ...`. Was there a reason behind that? Otherwise it would probably be best to keep the "diff" as small as possible and refrain from stylistic changes that don't touch the code you're modifying anyway. I'm myself a huge offender in that regard so perhaps that's why I noticed it. :sweat_smile: 

The whitespace fixes are a clear improvement though.

rfezzani replied:
I first wanted to make the imports respect the suggestion of [the python style guide](https://www.python.org/dev/peps/pep-0008/#imports), but I confess that I was tempted by the stylistic changes that you catched :smile:.
I can revert it if you wish.

rfezzani replied:
I first wanted to make the imports respect the suggestion of the python style guide, but I confess that I was tempted by the stylistic changes that you catched smile.
I can revert it if you wish.

jni replied:
I agree with @lagru. Best to keep diffs small. In some cases, it is worth making the changes, but then they should be done in their own PR. Later when one is doing `git bisect` to find when a thorny bug appeared, small diffs are extremely appreciated!

rfezzani replied:
I reverted the stylistic changes ;-)

lagru commented on skimage/feature/tests/test_match.py:
```diff
@@ -50,14 +50,17 @@ def test_binary_descriptors_rotation_crosscheck_false():
-                             13, 10,  9, 27,  7, 11, 15,  8, 23, 14, 12, 16,
-                             10, 25, 18, 19, 21, 20, 41, 24, 25, 26, 28, 27,
-                             22, 23, 29, 30, 31, 32, 35, 33, 34, 30, 36])
+    exp_matches1 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12,
+                             13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
+                             23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
+                             33, 34, 35, 36, 37, 38, 39, 40, 41, 42,
+                             43, 44, 45, 46])
```
```suggestion
    exp_matches1 = np.arange(47)
```

lagru commented on skimage/feature/corner.py:
```diff
@@ -917,16 +918,35 @@ def corner_subpix(image, corners, window_size=11, alpha=0.99):
+    image : ndarray
+        Input image.
+    min_distance : int, optional
+        The minimum distance seperating peaks. Use the ``p`` argument
+        to set the Minkowski p-norm defining the distance.
```
```suggestion
        The minimal allowed distance separating peaks. 
```
I don't think it's necessary to add a reference to the immediate next parameter. (And typo in sep**e**rating)

rfezzani commented:
Thank you @lagru for your review.

> Would you mind I quick explanation of why these changes change the peak order?

The peak's order is not modified by `corner_peak`, the problem pointed out by @emmanuelle comes directly from a [bug](https://github.com/scikit-image/scikit-image/pull/4218#issuecomment-612641502) in the  [_get_high_intensity_peaks](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/peak.py#L7) function.

stefanv commented on skimage/feature/corner.py:
```diff
@@ -917,16 +918,34 @@ def corner_subpix(image, corners, window_size=11, alpha=0.99):
-                 footprint=None, labels=None, *, num_peaks_per_label=np.inf):
+                 footprint=None, labels=None, *, num_peaks_per_label=np.inf,
+                 p_norm=np.inf):
     """Find corners in corner measure response image.
```
I know you haven't touched this, but I see now that it's a confusing phrasing.

How about: `Find peaks in corner response image, corresponding to corners in the original image.`

rfezzani replied:
Isn't `Find peaks in a corner response image.` clear enough?

stefanv commented on skimage/feature/corner.py:
```diff
@@ -917,16 +918,34 @@ def corner_subpix(image, corners, window_size=11, alpha=0.99):
+    p_norm : float
+        Which Minkowski p-norm to use. Should be in the range [1, inf].
+        A finite large p may cause a ValueError if overflow can occur.
+        inf corresponds to the Chebyshev distance and 2 to the
```
```suggestion
        ``inf`` corresponds to the Chebyshev distance and 2 to the
```

rfezzani commented:
@emmanuelle, I opened #4582 to fix the peaks' coordinates ordering.

rfezzani commented:
@emmanuelle, #4582 is now merged (thank you @jni and @sciunto ;-)), I fixed corners order in the doctests.

rfezzani commented on skimage/feature/corner.py:
```diff
@@ -956,38 +975,53 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
              "Until version 0.18, a None value corresponds to a threshold "
-             "value of 0.1. "
-             "The default behavior will match skimage.feature.peak_local_max. "
-             "To avoid this warning, set threshold_rel=0.",
```
Ooops! No, I will revert this, sorry :pray:

rfezzani commented:
Thank you @lagru for your review, I fixed the doctests!

rfezzani commented:
Thank you @lagru :tada:

emmanuelle commented:
Merging, thank you @rfezzani !

rfezzani commented:
:tada: Thank you @emmanuelle
